### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by SaifAbuR3d**

### CODE REVIEW

File: .github/workflows/build-and-test.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    - '**.cs'
+    - '**.csproj'
+
+env:
+  DOTNET_VERSION: '6.0.401' # The .NET SDK version to use
+
+jobs:
+  build-and-test:
+
+    name: build-and-test-${{matrix.os}}
```
Comment (IMasriR365) [inline .github/workflows/build-and-test.yaml:17]: what does matrix.os return? and why do we need it in the step name?

File: .github/workflows/build-and-test.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        dotnet-version: ${{ env.DOTNET_VERSION }}
+
+    - name: Install dependencies
+      run: dotnet restore
+      
+    - name: Build
+      run: dotnet build --configuration Release --no-restore
+    
+    - name: Test
+      run: dotnet test --no-restore --verbosity normal
```
Comment (IMasriR365) [inline .github/workflows/build-and-test.yaml:37]: why do we need to specify verbosity?

File: ATB.BL/ATB.csproj [hunk 3]
```diff
@@ -7,4 +7,8 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="CsvHelper" Version="30.0.1" />
+  </ItemGroup>
```
Comment (IMasriR365) [inline ATB.BL/ATB.csproj:12]: why do we need this package here?

File: ATB.BL/DataAccess/CsvUtility.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            int currentRow = 1;
+
+            while (csvReader.Read())
+            {
+                var validationResult = InputFlightDataValidation.ValidateFlightData(csvReader);
+
+                if (validationResult.IsValid)
+                {
+                    // the below code is null-safe.
+                    var flightId = int.Parse(csvReader.GetField(0));
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/CsvUtility.cs:31]: can you gurantee that even when the code for this method is changed?

File: ATB.BL/DataAccess/CsvUtility.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
+﻿using System.Globalization;
+using System.Text;
+using ATB.Entities;
+using CsvHelper;
+using CsvHelper.Configuration;
+
+namespace ATB.DataAccess
+{
+    internal static class CsvUtility
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/CsvUtility.cs:9]: what does this class do?

File: ATB.BL/DataAccess/CsvUtility.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                {
+                    foreach (Flight flight in flights)
+                    {
+                        string flightLine = $"{flight.FlightId},{flight.Price},{flight.DepartureCountry},{flight.DestinationCountry}," +
+                                            $"{flight.DepartureDate},{flight.DepartureAirport},{flight.ArrivalAirport},{flight.FClass}";
+
+                        writer.WriteLine(flightLine);
+                    }
+
+                    Console.WriteLine("Flights added successfully!");
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/CsvUtility.cs:106]: why would we need this?

File: ATB.BL/DataAccess/FileBookingRepository.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            try
+            {
+                string bookingLine = $"{booking.Passenger.PassengerId},{booking.Flight.FlightId},{booking.FClass}";
+                File.AppendAllLines(bookingsFilePath, new[] { bookingLine });
+                Console.WriteLine("Booking added successfully!");
+            }
+            catch (IOException ex)
+            {
+                Console.WriteLine($"An error occurred while adding the booking: {ex.Message}");
+            }
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FileBookingRepository.cs:20]: I think this was from a previous PR, but do we want to catch this exception here?

File: ATB.BL/DataAccess/FileBookingRepository.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        private void WriteBookingsIntoFile(IEnumerable<Booking> bookings) // OVERWRITES Existing Lines
+        {
+            try
+            {
+                using (StreamWriter writer = new StreamWriter(bookingsFilePath, false)) // Open the file in write mode, which overwrites existing content
+                {
+                    foreach (Booking booking in bookings)
+                    {
+                        string line = $"{booking.Passenger.PassengerId},{booking.Flight.FlightId},{booking.FClass}";
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FileBookingRepository.cs:68]: is this the responsibility of this class to format bookings?

File: ATB.BL/DataAccess/FileBookingRepository.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
+﻿using ATB.Entities;
+using ATB.Services;
+
+namespace ATB.DataAccess
+{
+    internal class FileBookingRepository : IBookingRepository
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FileBookingRepository.cs:6]: can you draw a small diagram showing how all these different classes interact?

File: ATB.BL/DataAccess/FileBookingRepository.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            {
+                Console.WriteLine($"An error occurred while writing to the file: {ex.Message}");
+            }
+        }
+        public void RemoveBooking(Booking booking)
+        {
+            // Contains All Bookings Except For (booking)
+            IEnumerable<Booking> updatedList = GetAllBookings().Where(_booking => !_booking.Equals(booking));
+
+            // Replace the contents of bookings.txt with the new list
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FileBookingRepository.cs:84]: what's the point of adding this comment?

File: ATB.BL/DataAccess/FileFlightRepository.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            return flightDictionary.Values;
+        }
+
+        public Flight? GetFlight(int flightId, FlightClass flightClass) // returns null if the Flight is not in the dictionary
+        {
+            if(flightDictionary.TryGetValue((flightId, flightClass), out Flight flight))
+            {
+                return flight;
+            }
+            return null;
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FileFlightRepository.cs:33]: could you do this without an if? if yes, how?

File: ATB.BL/DataAccess/FileFlightRepository.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+                        {
+                            FlightId = id,
+                            Price = price,
+                            DepartureCountry = departureCountry,
+                            DestinationCountry = destinationCountry,
+                            DepartureDate = departureDate,
+                            DepartureAirport = departureAirport,
+                            ArrivalAirport = arrivalAirport,
+                            FClass = fClass
+                        };
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FileFlightRepository.cs:67]: does this logic belong here?

File: ATB.BL/DataAccess/FilePassengerRepository.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+{
+    internal class FilePassengerRepository : IPassengerRepository
+    {
+        private string passengersFilePath = "files/Passengers.txt";
+
+        private Dictionary<int, Passenger> passengerDictionary;
+
+        public FilePassengerRepository()
+        {
+            InitializePassengerDictionary();
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/FilePassengerRepository.cs:13]: what's the point of separating this into a method?

File: ATB.BL/DataAccess/InputFlightDataValidation.cs [hunk 14]
```diff
@@ -0,0 +1,99 @@
+﻿using ATB.DataAccess;
+using ATB.Entities;
+using CsvHelper;
+
+internal static class InputFlightDataValidation
+{
+
+    internal static FlightValidationResult ValidateFlightData(CsvReader csvReader)
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/InputFlightDataValidation.cs:8]: how to decouple this method from the CSV library?

File: ATB.BL/DataAccess/InputFlightDataValidation.cs [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+        if (!Enum.TryParse<FlightClass>(csvReader.GetField(7), true, out _))
+        {
+            validationResult.Errors.Add($"Invalid FlightClass \"{csvReader.GetField(7) ?? String.Empty}\": Must be one of (Economy, Business, First Class)");
+        }
+
+        validationResult.IsValid = (validationResult.Errors.Count == 0);
+        return validationResult;
+    }
+
+    public static List<Flight> GetConsistentFlights(Dictionary<int, List<Flight>> flightGroups)
```
Comment (IMasriR365) [inline ATB.BL/DataAccess/InputFlightDataValidation.cs:55]: is this truly validation logic?

File: ATB.BL/Entities/FlightValidationResult.cs [hunk 16]
```diff
@@ -0,0 +1,8 @@
+﻿namespace ATB.Entities
+{
+    public class FlightValidationResult
+    {
+        public bool IsValid { get; set; }
```
Comment (IMasriR365) [inline ATB.BL/Entities/FlightValidationResult.cs:5]: can we make this a derived property?

File: ATB.BL/Entities/Manger.cs [hunk 17]
```diff
@@ -1,6 +1,8 @@
 ﻿namespace ATB.Entities
 {
-    internal class Manger : User
+
+    [Obsolete]
```
Comment (IMasriR365) [inline ATB.BL/Entities/Manger.cs:4]: why keep it?

File: ATB.BL/Helpers/UserInputHelper.cs [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ATB.Helpers
+{
+    public static class UserInputHelper
+    {
+
+        public static decimal GetValidDecimal(string promptMessage = "", string errorMessage = "Invalid input. Please enter a valid decimal.")
```
Comment (IMasriR365) [inline ATB.BL/Helpers/UserInputHelper.cs:13]: how do we combine all of those methods into one?

File: ATB.BL/Services/BookingService.cs [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+                     (booking.Flight.FClass.Equals(bookingSearchCriteria.FClass) || (bookingSearchCriteria.FClass is null)) &&
+                     (booking.Flight.DestinationCountry.Equals(bookingSearchCriteria.DestinationCountry) || (bookingSearchCriteria.DestinationCountry is null)) &&
+                     (booking.Flight.ArrivalAirport.Equals(bookingSearchCriteria.ArrivalAirport) || (bookingSearchCriteria.ArrivalAirport is null)) &&
+                     (booking.Flight.DepartureCountry.Equals(bookingSearchCriteria.DepartureCountry) || (bookingSearchCriteria.DepartureCountry is null)) &&
+                     (booking.Flight.DepartureAirport.Equals(bookingSearchCriteria.DepartureAirport) || (bookingSearchCriteria.DepartureAirport is null)) &&
+
+                     (booking.Flight.FlightId.Equals(bookingSearchCriteria.FlightId) || (bookingSearchCriteria.FlightId is null)) &&
+                     (booking.Passenger.PassengerId.Equals(bookingSearchCriteria.PassengerId) || (bookingSearchCriteria.PassengerId is null))
+                     );
```
Comment (IMasriR365) [inline ATB.BL/Services/BookingService.cs:33]: this looks a bit messy, how can I make a bit more clean?

File: ATB.BL/Presentation/Program.cs [hunk 20]
```diff
@@ -1,9 +1,443 @@
...
+using ATB.Entities;
+using ATB.Helpers;
+using ATB.Services;
+
+namespace ATB.Presentation
 {
-    internal class Program
-    {
-        static void Main(string[] args)
+    internal class Program
```
Comment (IMasriR365) [inline ATB.BL/Presentation/Program.cs:8]: this class is super bloated

File: ATB.Tests/BookingServiceTests.cs [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+{
+    public class BookingServiceTests
+    {
+        private readonly IFixture fixture;
+        private readonly Mock<IBookingRepository> mockBookingRepository;
+        private readonly BookingService sut;
+        public BookingServiceTests()
+        {
+            fixture = new Fixture().Customize(new AutoMoqCustomization());
+            mockBookingRepository = fixture.Freeze<Mock<IBookingRepository>>();
```
Comment (IMasriR365) [inline ATB.Tests/BookingServiceTests.cs:18]: what does freeze do?

File: ATB.Tests/BookingServiceTests.cs [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+
+        [Fact]
+        public void GetAllBookings_ReturnsAllBookings()
+        {
+            var bookings = fixture.CreateMany<Booking>();
+            mockBookingRepository.Setup(x => x.GetAllBookings()).Returns(bookings);
+
+            var result = sut.GetAllBookings();
+
+            Assert.Equal(bookings, result);
```
Comment (IMasriR365) [inline ATB.Tests/BookingServiceTests.cs:30]: what logic are we testing here?

File: ATB.Tests/BookingServiceTests.cs [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+        public void AddBooking_PassengerHasBookedThisFlightBefore_ReturnsFailed()
+        {
+            var passenger = fixture.Create<Passenger>();
+            var flight = fixture.Create<Flight>();
+            var flightClass = fixture.Create<FlightClass>();
+            var booking = new Booking(flight, passenger, flightClass);
+            var bookings = new List<Booking>
+            {
+                booking
+            };
```
Comment (IMasriR365) [inline ATB.Tests/BookingServiceTests.cs:61]: can we use auto fixture for those initialization?

File: ATB.Tests/BookingServiceTests.cs [hunk 24]
```diff
@@ -0,0 +1,297 @@
...
+            mockBookingRepository.Verify(x => x.UpdateBookingClass(booking, newFlightClass), Times.Never());
+        }
+        [Fact]
+        public void UpdateBookingClass_PassengerHasNotBookedThisFlightBefore_ReturnsFailed()
+        {
+            var passenger = fixture.Create<Passenger>();
+            var flight = fixture.Create<Flight>();
+            var oldFlightClass = fixture.Create<FlightClass>();
+
+            var booking = new Booking(flight, passenger, oldFlightClass);  // real booking
```
Comment (IMasriR365) [inline ATB.Tests/BookingServiceTests.cs:196]: what do you mean by real booking?

File: ATB.Tests/BookingServiceTests.cs [hunk 25]
```diff
@@ -0,0 +1,297 @@
...
+            var searchCriteria = new BookingSearchCriteria
+            {
+                Price = 100.0m
+            };
+
+            var result = sut.FilterBookings(searchCriteria);
+            Assert.Contains(matchingBooking, result);
+        }
+        [Fact]
+        public void FilterBookings_WithMultipleCriteriaAndMultipleMatchingBookings_ReturnsMatchingBookings()
```
Comment (IMasriR365) [inline ATB.Tests/BookingServiceTests.cs:253]: does this cover enough?

File: ATB.Tests/FlightServiceTests.cs [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+
+            var result = sut.GetAllFlights();
+
+            Assert.Equal(flights, result);
+        }
+
+
+
+
+        public void FilterFlights_WithNullCriteria_ReturnsAllFlights()
```
Comment (IMasriR365) [inline ATB.Tests/FlightServiceTests.cs:35]: no [Fact]?

File: ATB.BL/DataAccess/CsvUtility.cs (hunk 4 above)
Comment (SaifAbuR3d) [reply ATB.BL/DataAccess/CsvUtility.cs:31]: if (csvReader is not null) {
  var flightId = int.TryParse(csvReader.GetField(0), out var parsedFlightId) ? parsedFlightId : 0;
  var price = decimal.TryParse(csvReader.GetField(1), out var parsedPrice) ? parsedPrice : 0;
  var departureCountry = csvReader.GetField(2) ?? string.Empty;
  // ....
  var fClass = Enum.TryParse<FlightClass>(csvReader.GetField(7), true, out var parsedFClass) ? parsedFClass : 
  FlightClass.Economy;
}

File: ATB.BL/DataAccess/FileFlightRepository.cs (hunk 11 above)
Comment (SaifAbuR3d) [reply ATB.BL/DataAccess/FileFlightRepository.cs:33]: return flightDictionary.TryGetValue((flightId, flightClass), out var flight) ? flight : null;

File: ATB.BL/DataAccess/FileFlightRepository.cs (hunk 12 above)
Comment (SaifAbuR3d) [reply ATB.BL/DataAccess/FileFlightRepository.cs:67]: refactored code (in Dev branch) 
 private void InitializeFlightDictionary()
        {
            flightDictionary = new Dictionary<(int, FlightClass), Flight>();

            try
            {
                string[] lines = File.ReadAllLines(flightsFilePath);

                foreach (string line in lines)
                {
                    string[] values = line.Split(',');
                    **var flight = FlightParser.ParseFlightFromStrings(values);**
                    flightDictionary[(flight.FlightId, flight.FClass)] = flight;
                }
            }
            catch (IOException ex)
            {
                Console.WriteLine($"An error occurred while reading flights from the file: {ex.Message}");
            }
        }

File: ATB.BL/Services/BookingService.cs (hunk 19 above)
Comment (SaifAbuR3d) [reply ATB.BL/Services/BookingService.cs:33]: refactored code #1 

 public IEnumerable<Booking> FilterBookings(BookingSearchCriteria bookingSearchCriteria)
        {
            var allBookings = GetAllBookings();

            bool IsNotSpecified<T>(T value) => value == null;

            bool FilterPredicate(Booking booking)
            {
                var flight = booking.Flight;

                bool IsPriceMatch = flight.Price.Equals(bookingSearchCriteria.Price) || IsNotSpecified(bookingSearchCriteria.Price);
                bool IsDepartureDateMatch = flight.DepartureDate.Equals(bookingSearchCriteria.DepartureDate) || IsNotSpecified(bookingSearchCriteria.DepartureDate);
                bool IsFClassMatch = flight.FClass.Equals(bookingSearchCriteria.FClass) || IsNotSpecified(bookingSearchCriteria.FClass);
                bool IsDestinationCountryMatch = flight.DestinationCountry.Equals(bookingSearchCriteria.DestinationCountry) || IsNotSpecified(bookingSearchCriteria.DestinationCountry);
                bool IsArrivalAirportMatch = flight.ArrivalAirport.Equals(bookingSearchCriteria.ArrivalAirport) || IsNotSpecified(bookingSearchCriteria.ArrivalAirport);
                bool IsDepartureCountryMatch = flight.DepartureCountry.Equals(bookingSearchCriteria.DepartureCountry) || IsNotSpecified(bookingSearchCriteria.DepartureCountry);
                bool IsDepartureAirportMatch = flight.DepartureAirport.Equals(bookingSearchCriteria.DepartureAirport) || IsNotSpecified(bookingSearchCriteria.DepartureAirport);
                bool IsFlightIdMatch = flight.FlightId.Equals(bookingSearchCriteria.FlightId) || IsNotSpecified(bookingSearchCriteria.FlightId);
                bool IsPassengerIdMatch = booking.Passenger.PassengerId.Equals(bookingSearchCriteria.PassengerId) || IsNotSpecified(bookingSearchCriteria.PassengerId);

                return IsPriceMatch && IsDepartureDateMatch && IsFClassMatch &&
                       IsDestinationCountryMatch && IsArrivalAirportMatch &&
                       IsDepartureCountryMatch && IsDepartureAirportMatch &&
                       IsFlightIdMatch && IsPassengerIdMatch;
            }

            var filteredBookings = allBookings.Where(FilterPredicate);

            return filteredBookings;
        }

File: ATB.BL/Entities/FlightValidationResult.cs (hunk 16 above)
Comment (SaifAbuR3d) [reply ATB.BL/Entities/FlightValidationResult.cs:5]: namespace ATB.Entities
{
    public class FlightValidationResult
    {
        private List<string> _errors = new List<string>();

        public bool IsValid => _errors.Count == 0;

        public List<string> Errors => _errors;
    }
}

File: ATB.Tests/BookingServiceTests.cs (hunk 23 above)
Comment (SaifAbuR3d) [reply ATB.Tests/BookingServiceTests.cs:61]: var booking = fixture.Build<Booking>()
                .With(b => b.Flight, flight)
                .With(b => b.Passenger, passenger)
                .With(b => b.FClass, flightClass)
                .Create();

File: ATB.BL/DataAccess/InputFlightDataValidation.cs (hunk 15 above)
Comment (IMasriR365) [reply ATB.BL/DataAccess/InputFlightDataValidation.cs:55]: how to apply CQS?

**Pull request merged**